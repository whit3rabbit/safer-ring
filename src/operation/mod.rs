//! Type-safe operation state management.
//!
//! This module provides a type-safe state machine for io_uring operations that prevents
//! common usage errors at compile time. Operations progress through well-defined states:
//!
//! - [`Building`]: Initial state where the operation can be configured
//! - [`Submitted`]: Operation has been submitted to the kernel and is in flight
//! - [`Completed<T>`]: Operation has finished and contains the result
//!
//! State transitions are enforced at compile time, making it impossible to:
//! - Submit an operation twice
//! - Poll an operation that hasn't been submitted
//! - Extract results from an incomplete operation
//!
//! # Example
//!
//! ```rust,no_run
//! # use safer_ring::operation::{Operation, OperationType};
//! # use std::pin::Pin;
//! let mut buffer = vec![0u8; 1024];
//! let pinned = Pin::new(buffer.as_mut_slice());
//!
//! let operation = Operation::read()
//!     .fd(0)
//!     .buffer(pinned)
//!     .offset(0);
//! ```

// Public modules
pub mod types;

// Internal modules
mod building;
mod completed;
mod core;
mod states;
mod submitted;

// Internal operation tracking module
pub(crate) mod tracker;

// Test module
#[cfg(test)]
mod tests;

// Re-exports for public API
pub use core::{BufferType, FdType, Operation};
pub use states::{Building, Completed, OperationState, Submitted};
pub use types::OperationType;
