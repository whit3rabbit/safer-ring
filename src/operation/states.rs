//! Operation state types and state machine definitions.
//!
//! This module defines the state types used in the Operation state machine
//! and the sealed OperationState trait that ensures type safety.

/// Marker trait for operation states.
///
/// This trait is sealed to prevent external implementations and ensure
/// that only valid states can be used with operations.
pub trait OperationState: private::Sealed {}

/// Operation is being built and can be configured.
///
/// In this state, the operation can have its parameters modified but cannot
/// be submitted to the kernel. This prevents incomplete operations from being
/// submitted accidentally.
#[derive(Debug, Clone, Copy)]
pub struct Building;

/// Operation has been submitted and is in flight.
///
/// In this state, the operation cannot be modified but can be polled for
/// completion. The operation ID is used to track the operation in the kernel.
#[derive(Debug, Clone, Copy)]
pub struct Submitted {
    /// Operation ID for tracking in the completion queue
    /// This ID is assigned by the ring when the operation is submitted
    pub(crate) id: u64,
}

/// Operation has completed with a result.
///
/// In this state, the operation result can be extracted along with the
/// buffer ownership. This is the terminal state for operations.
#[derive(Debug)]
pub struct Completed<T> {
    /// The result of the operation
    pub(crate) result: T,
}

// Implement the sealed trait for all valid states
impl OperationState for Building {}
impl OperationState for Submitted {}
impl<T> OperationState for Completed<T> {}

/// Private module to seal the OperationState trait.
///
/// This prevents external crates from implementing OperationState,
/// ensuring type safety of the state machine.
mod private {
    pub trait Sealed {}
    impl Sealed for super::Building {}
    impl Sealed for super::Submitted {}
    impl<T> Sealed for super::Completed<T> {}
}
