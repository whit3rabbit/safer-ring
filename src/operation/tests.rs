//! Comprehensive tests for operation state management.

use super::*;
use std::pin::Pin;

// Compile-time assertions to ensure our types have the expected properties
#[cfg(test)]
mod compile_time_tests {
    use super::*;
    use static_assertions::*;

    // Ensure operations are Send when appropriate
    assert_impl_all!(Operation<'static, 'static, Building>: Send);
    assert_impl_all!(Operation<'static, 'static, Submitted>: Send);
    assert_impl_all!(Operation<'static, 'static, Completed<i32>>: Send);

    // Ensure state types are the expected size for optimal memory layout
    assert_eq_size!(Building, ()); // Zero-sized type
    assert_eq_size!(Submitted, u64); // Just the operation ID
}

#[test]
fn test_operation_creation() {
    let op = Operation::<'_, '_, Building>::new();
    assert_eq!(op.get_fd(), -1);
    assert_eq!(op.get_offset(), 0);
    assert_eq!(op.get_type(), OperationType::Read);
    assert!(!op.has_buffer());
}

#[test]
fn test_operation_builder_methods() {
    let op = Operation::read().fd(5).offset(1024);

    assert_eq!(op.get_fd(), 5);
    assert_eq!(op.get_offset(), 1024);
    assert_eq!(op.get_type(), OperationType::Read);
}

#[test]
fn test_operation_types() {
    assert_eq!(Operation::read().get_type(), OperationType::Read);
    assert_eq!(Operation::write().get_type(), OperationType::Write);
    assert_eq!(Operation::accept().get_type(), OperationType::Accept);
    assert_eq!(Operation::send().get_type(), OperationType::Send);
    assert_eq!(Operation::recv().get_type(), OperationType::Recv);
}

#[test]
fn test_operation_validation() {
    // Invalid: no fd set
    let op = Operation::read();
    assert!(op.validate().is_err());

    // Invalid: fd set but no buffer for I/O operation
    let op = Operation::read().fd(0);
    assert!(op.validate().is_err());

    // Valid: accept operation doesn't need buffer
    let op = Operation::accept().fd(0);
    assert!(op.validate().is_ok());
}

#[test]
fn test_operation_with_buffer() {
    let mut buffer = vec![0u8; 1024];
    let pinned = Pin::new(buffer.as_mut_slice());

    let op = Operation::read().fd(0).buffer(pinned);

    assert!(op.has_buffer());
    assert!(op.validate().is_ok());
}

#[test]
fn test_state_transitions() {
    let mut buffer = vec![0u8; 1024];
    let pinned = Pin::new(buffer.as_mut_slice());

    let op = Operation::read().fd(0).buffer(pinned);

    // Building -> Submitted
    let submitted = op.submit_with_id(42).unwrap();
    assert_eq!(submitted.id(), 42);
    assert_eq!(submitted.fd(), 0);

    // Submitted -> Completed
    let completed = submitted.complete_with_result(100i32);
    assert_eq!(*completed.result(), 100);

    // Completed -> extracted
    let (result, _buffer) = completed.into_result();
    assert_eq!(result, 100);
}

#[test]
fn test_operation_properties() {
    let mut buffer = vec![0u8; 1024];
    let pinned = Pin::new(buffer.as_mut_slice());

    let submitted = Operation::write()
        .fd(1)
        .offset(512)
        .buffer(pinned)
        .submit_with_id(123)
        .unwrap();

    assert_eq!(submitted.id(), 123);
    assert_eq!(submitted.fd(), 1);
    assert_eq!(submitted.offset(), 512);
    assert_eq!(submitted.op_type(), OperationType::Write);
}

#[test]
fn test_completed_operation_access() {
    let mut buffer = vec![0u8; 1024];
    let pinned = Pin::new(buffer.as_mut_slice());

    let completed = Operation::read()
        .fd(0)
        .buffer(pinned)
        .submit_with_id(1)
        .unwrap()
        .complete_with_result("test result".to_string());

    // Can access result by reference
    assert_eq!(completed.result(), "test result");
    assert_eq!(completed.fd(), 0);
    assert_eq!(completed.op_type(), OperationType::Read);

    // Can extract result by value
    let (result, _buffer) = completed.into_result();
    assert_eq!(result, "test result");
}

#[test]
fn test_operation_state_machine_invariants() {
    let mut buffer = vec![0u8; 1024];
    let pinned = Pin::new(buffer.as_mut_slice());

    let building = Operation::read().fd(0).buffer(pinned);

    // Building -> Submitted
    let submitted = building.submit_with_id(1).unwrap();
    assert_eq!(submitted.id(), 1);

    // Submitted -> Completed
    let completed = submitted.complete_with_result(42);
    assert_eq!(*completed.result(), 42);

    // Completed -> extracted
    let (result, _) = completed.into_result();
    assert_eq!(result, 42);
}

#[test]
fn test_operation_validation_rules() {
    // Missing FD
    let op = Operation::read();
    assert!(op.validate().is_err());

    // Missing buffer for I/O operations
    let op = Operation::read().fd(0);
    assert!(op.validate().is_err());

    let op = Operation::write().fd(0);
    assert!(op.validate().is_err());

    // Accept doesn't need buffer
    let op = Operation::accept().fd(0);
    assert!(op.validate().is_ok());
}

#[test]
fn test_builder_pattern_fluency() {
    // Test that the builder pattern works fluently
    let op = Operation::write().fd(1).offset(1024);

    assert_eq!(op.get_fd(), 1);
    assert_eq!(op.get_offset(), 1024);
    assert_eq!(op.get_type(), OperationType::Write);
}

#[test]
fn test_default_implementation() {
    let op1 = Operation::<'_, '_, Building>::default();
    let op2 = Operation::<'_, '_, Building>::new();

    // Both should have the same initial state
    assert_eq!(op1.get_fd(), op2.get_fd());
    assert_eq!(op1.get_offset(), op2.get_offset());
    assert_eq!(op1.get_type(), op2.get_type());
    assert_eq!(op1.has_buffer(), op2.has_buffer());
}
