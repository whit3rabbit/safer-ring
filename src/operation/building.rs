//! Building state implementation and operation constructors.
//!
//! This module contains all the functionality for operations in the Building state,
//! including constructors and configuration methods.

use std::marker::PhantomData;
use std::os::unix::io::RawFd;
use std::pin::Pin;

use super::core::{BufferType, FdType, Operation};
use crate::operation::{Building, OperationType, Submitted};
use crate::registry::{RegisteredBuffer, RegisteredFd};

impl<'ring, 'buf> Default for Operation<'ring, 'buf, Building> {
    fn default() -> Self {
        Self::new()
    }
}

impl<'ring, 'buf> Operation<'ring, 'buf, Building> {
    /// Create a new operation in the building state.
    ///
    /// The operation starts with default values and must be configured
    /// before it can be submitted.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// let op = Operation::new()
    ///     .fd(0)
    ///     .offset(1024);
    /// ```
    pub fn new() -> Self {
        Self::with_type(OperationType::Read) // Default to read for backwards compatibility
    }

    /// Create a new read operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer = vec![0u8; 1024];
    /// let op = Operation::read()
    ///     .fd(0)
    ///     .buffer(Pin::new(buffer.as_mut_slice()));
    /// ```
    #[inline]
    pub fn read() -> Self {
        Self::with_type(OperationType::Read)
    }

    /// Create a new vectored read operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer1 = vec![0u8; 512];
    /// let mut buffer2 = vec![0u8; 512];
    /// let buffers = vec![
    ///     Pin::new(buffer1.as_mut_slice()),
    ///     Pin::new(buffer2.as_mut_slice()),
    /// ];
    /// let op = Operation::read_vectored()
    ///     .fd(0)
    ///     .buffers(buffers);
    /// ```
    #[inline]
    pub fn read_vectored() -> Self {
        Self::with_type(OperationType::ReadVectored)
    }

    /// Create a new write operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer = b"Hello, world!".to_vec();
    /// let op = Operation::write()
    ///     .fd(1)
    ///     .buffer(Pin::new(buffer.as_mut_slice()));
    /// ```
    #[inline]
    pub fn write() -> Self {
        Self::with_type(OperationType::Write)
    }

    /// Create a new vectored write operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer1 = b"Hello, ".to_vec();
    /// let mut buffer2 = b"world!".to_vec();
    /// let buffers = vec![
    ///     Pin::new(buffer1.as_mut_slice()),
    ///     Pin::new(buffer2.as_mut_slice()),
    /// ];
    /// let op = Operation::write_vectored()
    ///     .fd(1)
    ///     .buffers(buffers);
    /// ```
    #[inline]
    pub fn write_vectored() -> Self {
        Self::with_type(OperationType::WriteVectored)
    }

    /// Create a new accept operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// let op = Operation::accept().fd(3); // listening socket fd
    /// ```
    #[inline]
    pub fn accept() -> Self {
        Self::with_type(OperationType::Accept)
    }

    /// Create a new send operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer = b"Hello, client!".to_vec();
    /// let op = Operation::send()
    ///     .fd(4)
    ///     .buffer(Pin::new(buffer.as_mut_slice()));
    /// ```
    #[inline]
    pub fn send() -> Self {
        Self::with_type(OperationType::Send)
    }

    /// Create a new receive operation.
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer = vec![0u8; 1024];
    /// let op = Operation::recv()
    ///     .fd(4)
    ///     .buffer(Pin::new(buffer.as_mut_slice()));
    /// ```
    #[inline]
    pub fn recv() -> Self {
        Self::with_type(OperationType::Recv)
    }

    /// Create a new operation with the specified type.
    ///
    /// This is a helper method to reduce code duplication in the constructor methods.
    #[inline]
    fn with_type(op_type: OperationType) -> Self {
        Self {
            ring: PhantomData,
            buffer: BufferType::None,
            fd: FdType::Raw(-1), // Invalid fd that must be set before submission
            offset: 0,
            op_type,
            state: Building,
        }
    }

    /// Set the file descriptor for this operation.
    ///
    /// This is required for all operations and must be a valid file descriptor.
    ///
    /// # Arguments
    ///
    /// * `fd` - A valid file descriptor (>= 0)
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// let op = Operation::read().fd(0); // stdin
    /// ```
    #[inline]
    pub fn fd(mut self, fd: RawFd) -> Self {
        self.fd = FdType::Raw(fd);
        self
    }

    /// Set a registered file descriptor for this operation.
    ///
    /// Using registered file descriptors can improve performance for frequently
    /// used file descriptors by avoiding kernel lookups.
    ///
    /// # Arguments
    ///
    /// * `registered_fd` - A registered file descriptor
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::{operation::Operation, Registry};
    /// # fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let mut registry = Registry::new();
    /// let registered_fd = registry.register_fd(0)?;
    /// let op = Operation::read().registered_fd(registered_fd);
    /// # Ok(())
    /// # }
    /// ```
    #[inline]
    pub fn registered_fd(mut self, registered_fd: RegisteredFd) -> Self {
        self.fd = FdType::Registered(registered_fd);
        self
    }

    /// Set the buffer for this operation.
    ///
    /// The buffer lifetime must be at least as long as the ring lifetime
    /// to ensure memory safety during the operation. The buffer must remain
    /// pinned in memory until the operation completes.
    ///
    /// # Arguments
    ///
    /// * `buffer` - A pinned mutable slice that will be used for I/O
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut data = vec![0u8; 4096];
    /// let op = Operation::read()
    ///     .fd(0)
    ///     .buffer(Pin::new(data.as_mut_slice()));
    /// ```
    #[inline]
    pub fn buffer(mut self, buffer: Pin<&'buf mut [u8]>) -> Self {
        self.buffer = BufferType::Pinned(buffer);
        self
    }

    /// Set a registered buffer for this operation.
    ///
    /// Using registered buffers can improve performance by avoiding kernel
    /// buffer validation and setup overhead.
    ///
    /// # Arguments
    ///
    /// * `registered_buffer` - A registered buffer
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::{operation::Operation, Registry};
    /// # use std::pin::Pin;
    /// # fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let mut registry = Registry::new();
    /// let buffer = Pin::new(Box::new([0u8; 1024]));
    /// let registered_buffer = registry.register_buffer(buffer)?;
    /// let op = Operation::read()
    ///     .fd(0)
    ///     .registered_buffer(registered_buffer);
    /// # Ok(())
    /// # }
    /// ```
    #[inline]
    pub fn registered_buffer(mut self, registered_buffer: RegisteredBuffer) -> Self {
        self.buffer = BufferType::Registered(registered_buffer);
        self
    }

    /// Set multiple buffers for vectored I/O operations.
    ///
    /// This enables scatter-gather I/O where data can be read into or written
    /// from multiple non-contiguous buffers in a single operation.
    ///
    /// # Arguments
    ///
    /// * `buffers` - A vector of pinned mutable slices
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # use std::pin::Pin;
    /// let mut buffer1 = vec![0u8; 512];
    /// let mut buffer2 = vec![0u8; 512];
    /// let buffers = vec![
    ///     Pin::new(buffer1.as_mut_slice()),
    ///     Pin::new(buffer2.as_mut_slice()),
    /// ];
    /// let op = Operation::read_vectored()
    ///     .fd(0)
    ///     .buffers(buffers);
    /// ```
    #[inline]
    pub fn buffers(mut self, buffers: Vec<Pin<&'buf mut [u8]>>) -> Self {
        self.buffer = BufferType::Vectored(buffers);
        self
    }

    /// Set the offset for this operation.
    ///
    /// This is used for file operations to specify the position in the file.
    /// For socket operations, this parameter is typically ignored by the kernel.
    ///
    /// # Arguments
    ///
    /// * `offset` - Byte offset in the file (0-based)
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// let op = Operation::read()
    ///     .fd(3)
    ///     .offset(1024); // Read starting at byte 1024
    /// ```
    #[inline]
    pub fn offset(mut self, offset: u64) -> Self {
        self.offset = offset;
        self
    }



    /// Get the current offset.
    ///
    /// Returns the byte offset for file operations, or 0 if not set.
    #[inline]
    pub fn get_offset(&self) -> u64 {
        self.offset
    }

    /// Validate that the operation is ready for submission.
    ///
    /// This checks that all required fields are set for the operation type.
    /// Using the type system's knowledge about operation requirements for
    /// more efficient validation.
    ///
    /// # Errors
    ///
    /// Returns an error if:
    /// - File descriptor is not set (< 0)
    /// - Buffer is required but not set
    /// - Vectored operation has empty buffer list
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// let op = Operation::read().fd(0);
    /// assert!(op.validate().is_err()); // Missing buffer
    /// ```
    pub fn validate(&self) -> Result<(), &'static str> {
        // Check file descriptor
        match &self.fd {
            FdType::Raw(fd) if *fd < 0 => return Err("File descriptor must be set"),
            _ => {}
        }

        // Use the type system's knowledge about buffer requirements
        if self.op_type.requires_buffer() {
            match &self.buffer {
                BufferType::None => return Err("Buffer must be set for I/O operations"),
                BufferType::Vectored(buffers) if buffers.is_empty() => {
                    return Err("Vectored operations require at least one buffer")
                }
                _ => {}
            }
        }

        // Validate operation type matches buffer type
        if self.op_type.is_vectored() && !matches!(self.buffer, BufferType::Vectored(_)) {
            return Err("Vectored operation types require vectored buffers");
        }

        if !self.op_type.is_vectored() && matches!(self.buffer, BufferType::Vectored(_)) {
            return Err("Non-vectored operation types cannot use vectored buffers");
        }

        Ok(())
    }

    /// Convert this building operation to a submitted operation.
    ///
    /// This is typically called by the Ring when submitting the operation.
    /// It validates the operation and transitions to the Submitted state.
    ///
    /// # Arguments
    ///
    /// * `id` - Unique operation ID assigned by the ring
    ///
    /// # Errors
    ///
    /// Returns an error if the operation is not properly configured.
    #[allow(dead_code)]
    pub(crate) fn submit_with_id(
        self,
        id: u64,
    ) -> Result<Operation<'ring, 'buf, Submitted>, &'static str> {
        self.validate()?;

        // Zero-cost state transition - just change the type parameter
        Ok(Operation {
            ring: self.ring,
            buffer: self.buffer,
            fd: self.fd,
            offset: self.offset,
            op_type: self.op_type,
            state: Submitted { id },
        })
    }
}
