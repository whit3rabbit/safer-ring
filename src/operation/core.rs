//! Core Operation struct definition and common functionality.
//!
//! This module contains the main Operation struct and methods that are
//! available across all states.

use std::marker::PhantomData;
use std::os::unix::io::RawFd;
use std::pin::Pin;

use crate::operation::{OperationState, OperationType};
use crate::registry::{RegisteredBuffer, RegisteredFd};

/// Buffer type for operations.
///
/// This enum represents the different types of buffers that can be used
/// with operations, supporting both regular pinned buffers and registered buffers.
#[derive(Debug)]
pub enum BufferType<'buf> {
    /// No buffer (for operations like accept)
    None,
    /// Regular pinned buffer
    Pinned(Pin<&'buf mut [u8]>),
    /// Registered buffer with the registry
    Registered(RegisteredBuffer),
    /// Multiple pinned buffers for vectored I/O
    Vectored(Vec<Pin<&'buf mut [u8]>>),
}

/// File descriptor type for operations.
///
/// This enum represents the different types of file descriptors that can be used
/// with operations, supporting both regular file descriptors and registered ones.
#[derive(Debug)]
pub enum FdType {
    /// Regular file descriptor
    Raw(RawFd),
    /// Registered file descriptor
    Registered(RegisteredFd),
}

/// Type-safe operation with compile-time state tracking.
///
/// The operation struct uses phantom types to track its current state and enforce
/// valid state transitions at compile time. The lifetime parameters ensure that:
/// - `'ring`: The operation cannot outlive the ring that will execute it
/// - `'buf`: The buffer cannot be dropped while the operation is in flight
///
/// # Type Parameters
///
/// - `'ring`: Lifetime of the ring that will execute this operation
/// - `'buf`: Lifetime of the buffer used by this operation
/// - `S`: Current state of the operation ([`crate::operation::Building`], [`crate::operation::Submitted`], or [`crate::operation::Completed<T>`])
#[derive(Debug)]
pub struct Operation<'ring, 'buf, S> {
    // PhantomData is zero-sized, so this doesn't add runtime overhead
    pub(crate) ring: PhantomData<&'ring ()>,
    // Buffer type supports different buffer configurations
    pub(crate) buffer: BufferType<'buf>,
    // File descriptor type supports both raw and registered fds
    pub(crate) fd: FdType,
    // u64 offset supports large files (>4GB) on all platforms
    pub(crate) offset: u64,
    pub(crate) op_type: OperationType,
    pub(crate) state: S,
}

impl<'ring, 'buf, S: OperationState> Operation<'ring, 'buf, S> {
    /// Get the operation type.
    ///
    /// Returns the type of I/O operation this will perform.
    #[inline]
    pub fn get_type(&self) -> OperationType {
        self.op_type
    }

    /// Check if the operation has a buffer set.
    ///
    /// Returns `true` if a buffer has been configured for this operation.
    #[inline]
    pub fn has_buffer(&self) -> bool {
        !matches!(self.buffer, BufferType::None)
    }

    /// Get the file descriptor for this operation.
    ///
    /// Returns the raw file descriptor, resolving registered fds if necessary.
    #[inline]
    pub fn get_fd(&self) -> RawFd {
        match &self.fd {
            FdType::Raw(fd) => *fd,
            FdType::Registered(reg_fd) => reg_fd.raw_fd(),
        }
    }


}
