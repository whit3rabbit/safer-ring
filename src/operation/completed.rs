//! Completed state implementation for finished operations.
//!
//! This module contains functionality for operations that have completed
//! and contain their results.

use std::os::unix::io::RawFd;
use std::pin::Pin;

use super::core::{BufferType, FdType, Operation};
use crate::operation::{Completed, OperationType};

impl<'ring, 'buf, T> Operation<'ring, 'buf, Completed<T>> {
    /// Extract the result from the completed operation.
    ///
    /// This consumes the operation and returns both the result and the buffer
    /// ownership, ensuring no resource leaks. This is the primary way to
    /// retrieve results from completed operations.
    ///
    /// # Returns
    ///
    /// A tuple containing:
    /// - The operation result
    /// - The buffer (if one was used), returned to the caller for reuse
    ///
    /// # Example
    ///
    /// ```rust,no_run
    /// # use safer_ring::operation::Operation;
    /// # let completed_op: Operation<'_, '_, safer_ring::operation::Completed<i32>> = todo!();
    /// let (bytes_read, buffer) = completed_op.into_result();
    /// println!("Read {} bytes", bytes_read);
    /// // Buffer can now be reused for another operation
    /// ```
    pub fn into_result(self) -> (T, BufferType<'buf>) {
        (self.state.result, self.buffer)
    }

    /// Extract the result and return a pinned buffer if available.
    ///
    /// This is a convenience method for operations that use regular pinned buffers.
    /// For other buffer types, use `into_result()` instead.
    pub fn into_result_with_buffer(self) -> (T, Option<Pin<&'buf mut [u8]>>) {
        let (result, buffer_type) = self.into_result();
        let buffer = match buffer_type {
            BufferType::Pinned(buf) => Some(buf),
            _ => None,
        };
        (result, buffer)
    }

    /// Extract the result and return vectored buffers if available.
    ///
    /// This is a convenience method for vectored operations.
    pub fn into_result_with_vectored_buffers(self) -> (T, Option<Vec<Pin<&'buf mut [u8]>>>) {
        let (result, buffer_type) = self.into_result();
        let buffers = match buffer_type {
            BufferType::Vectored(bufs) => Some(bufs),
            _ => None,
        };
        (result, buffers)
    }

    /// Get a reference to the result without consuming the operation.
    ///
    /// This allows inspecting the result while keeping the operation intact.
    #[inline]
    pub fn result(&self) -> &T {
        &self.state.result
    }

    /// Get the file descriptor for this operation.
    #[inline]
    pub fn fd(&self) -> RawFd {
        match &self.fd {
            FdType::Raw(fd) => *fd,
            FdType::Registered(reg_fd) => reg_fd.raw_fd(),
        }
    }

    /// Get the offset for this operation.
    #[inline]
    pub fn offset(&self) -> u64 {
        self.offset
    }

    /// Get the operation type.
    #[inline]
    pub fn op_type(&self) -> OperationType {
        self.op_type
    }
}
