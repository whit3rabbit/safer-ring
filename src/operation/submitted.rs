//! Submitted state implementation for in-flight operations.
//!
//! This module contains functionality for operations that have been submitted
//! to the kernel and are currently in flight.

use std::os::unix::io::RawFd;

use super::core::{BufferType, FdType, Operation};
use crate::operation::{Completed, OperationType, Submitted};

impl<'ring, 'buf> Operation<'ring, 'buf, Submitted> {
    /// Get the operation ID.
    ///
    /// This ID is used to track the operation in the completion queue
    /// and match completions to their corresponding operations.
    #[inline]
    pub fn id(&self) -> u64 {
        self.state.id
    }

    /// Get the file descriptor for this operation.
    #[inline]
    pub fn fd(&self) -> RawFd {
        match &self.fd {
            FdType::Raw(fd) => *fd,
            FdType::Registered(reg_fd) => reg_fd.raw_fd(),
        }
    }

    /// Get the offset for this operation.
    #[inline]
    pub fn offset(&self) -> u64 {
        self.offset
    }

    /// Get the operation type.
    #[inline]
    pub fn op_type(&self) -> OperationType {
        self.op_type
    }

    /// Get buffer information for kernel submission.
    ///
    /// Returns a tuple of (pointer, length) for the buffer if present.
    /// This is used internally by the Ring for kernel submission.
    ///
    /// # Safety
    ///
    /// The returned pointer is only valid while the buffer remains pinned
    /// and the operation is in flight. The caller must ensure the buffer
    /// is not moved or dropped until the operation completes.
    #[allow(dead_code)] // Used by ring submission logic
    pub(crate) fn buffer_info(&self) -> Option<(*mut u8, usize)> {
        match &self.buffer {
            BufferType::None => None,
            BufferType::Pinned(buf) => {
                let slice = buf.as_ref();
                Some((slice.as_ptr() as *mut u8, slice.len()))
            }
            BufferType::Registered(_) => {
                // For registered buffers, we need to get the buffer info from the registry
                // This will be handled by the ring during submission
                None
            }
            BufferType::Vectored(_) => {
                // For vectored operations, buffer info is handled differently
                // This will be handled by the ring during submission
                None
            }
        }
    }

    /// Get vectored buffer information for kernel submission.
    ///
    /// Returns a vector of (pointer, length) tuples for vectored operations.
    /// This is used internally by the Ring for vectored operation submission.
    ///
    /// # Safety
    ///
    /// The returned pointers are only valid while the buffers remain pinned
    /// and the operation is in flight.
    #[allow(dead_code)] // Used by ring submission logic
    pub(crate) fn vectored_buffer_info(&self) -> Option<Vec<(*mut u8, usize)>> {
        match &self.buffer {
            BufferType::Vectored(buffers) => {
                let info: Vec<_> = buffers
                    .iter()
                    .map(|buf| {
                        let slice = buf.as_ref();
                        (slice.as_ptr() as *mut u8, slice.len())
                    })
                    .collect();
                Some(info)
            }
            _ => None,
        }
    }

    /// Check if this operation uses a registered buffer.
    #[inline]
    pub fn uses_registered_buffer(&self) -> bool {
        matches!(self.buffer, BufferType::Registered(_))
    }

    /// Check if this operation uses a registered file descriptor.
    #[inline]
    pub fn uses_registered_fd(&self) -> bool {
        matches!(self.fd, FdType::Registered(_))
    }

    /// Check if this operation is vectored.
    #[inline]
    pub fn is_vectored(&self) -> bool {
        matches!(self.buffer, BufferType::Vectored(_))
    }

    /// Convert this submitted operation to a completed operation.
    ///
    /// This is typically called by the Ring when processing completions.
    /// The state transition is zero-cost at runtime.
    ///
    /// # Arguments
    ///
    /// * `result` - The result of the completed operation
    #[allow(dead_code)]
    pub(crate) fn complete_with_result<T>(self, result: T) -> Operation<'ring, 'buf, Completed<T>> {
        // Zero-cost state transition - just change the type parameter
        Operation {
            ring: self.ring,
            buffer: self.buffer,
            fd: self.fd,
            offset: self.offset,
            op_type: self.op_type,
            state: Completed { result },
        }
    }
}
