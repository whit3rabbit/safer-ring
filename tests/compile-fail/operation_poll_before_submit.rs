// This test should fail to compile because building operations don't have id() method

use safer_ring::operation::Operation;

fn main() {
    let operation = Operation::read()
        .fd(0);
    
    // This should not compile - building operations don't have id() method
    let _id = operation.id();
}