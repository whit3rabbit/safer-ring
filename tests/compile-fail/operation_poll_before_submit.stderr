error[E0599]: no method named `id` found for struct `Operation<'_, '_, Building>` in the current scope
  --> tests/compile-fail/operation_poll_before_submit.rs:10:25
   |
10 |     let _id = operation.id();
   |                         ^^
   |
help: there is a method `fd` with a similar name, but with different arguments
  --> src/operation/building.rs
   |
   |     pub fn fd(mut self, fd: RawFd) -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
